### ISSUE TITLE
Amount owed is lost when withdraw is made post depletion timestamp

### ISSUE THREAD

**Issue by smol-ninja** [sablier-labs/flow#125]
## tl;dr

When `lastTimeUpdate < depletionTime < block.timestamp`, in order words `remainingAmount < balance <  remainingAmount + _streamedAmountOf`, the implementation loses track of an amount that is equal to `streamedAmount - balance`. Since this amount is owed to the recipient, this is a serious issue.

## Step by step

Consider the following test scenario. Ideally, this should pass but in the current implementation, it would fail.

1. Sender changes rps before depletion time. This would update `remainingAmount` and `lastTimeUpdate`.
2. Let the time pass beyond depletion time.
3. Now the actual amount owed to the recipient = stored remaining amount + streamed amount since last time update. But since this would be > balance, if recipients withdraws max amount, he would receives the full balance. 
4. Now, the remaining owed amount to recipient should be = (ra + sa - bal) but if you fetch ra + sa from the API, it would return 0.
5. The remaining amount owed to recipient is lost.

<details><summary>Click to expand exploit test</summary>

```solidity
function test_Exploit() external {
    vm.warp({ newTimestamp: flow.depletionTimeOf(defaultStreamId) - 1000 });

    // Lets update `lastTimeUpdate` and  `remainingAmount` by calling `adjustRatePerSecond`.
    flow.adjustRatePerSecond(defaultStreamId, RATE_PER_SECOND - 1);
    // Lets quickly set it back to the original rate.
    flow.adjustRatePerSecond(defaultStreamId, RATE_PER_SECOND);

    // assertion: remainingAmount < balance and lastTimeUpdate < getDepletionTime
    assertLt(
        flow.getRemainingAmount(defaultStreamId), flow.getBalance(defaultStreamId), "remainingAmount < balance"
    );
    assertLt(
        flow.getLastTimeUpdate(defaultStreamId),
        flow.depletionTimeOf(defaultStreamId),
        "lastTimeUpdate < getDepletionTime"
    );

    vm.warp({ newTimestamp: flow.depletionTimeOf(defaultStreamId) + 1000 });

    // assertion: remainingAmount < balance <  remainingAmount + _streamedAmountOf
    assertLt(
        flow.getRemainingAmount(defaultStreamId), flow.getBalance(defaultStreamId), "remainingAmount < balance"
    );
    assertLt(
        flow.getBalance(defaultStreamId),
        flow.getRemainingAmount(defaultStreamId) + flow.streamedAmountOf(defaultStreamId),
        "balance < remainingAmount + _streamedAmountOf"
    );

    // Now if user withdraws, remainingAmount set to 0 , user should receive `balance` and there should be some
    // amount owed to the recipient (= remainingAmount + _streamedAmountOf - balance) which can be withdrawn after
    // next
    // deposit.
    resetPrank({ msgSender: users.recipient });
    vm.expectEmit({ emitter: address(dai) });
    emit IERC20.Transfer({
        from: address(flow),
        to: users.recipient,
        value: normalizeAmountWithStreamId(defaultStreamId, flow.getBalance(defaultStreamId))
    });
    flow.withdrawMax({ streamId: defaultStreamId, to: users.recipient });
    assertEq(flow.getRemainingAmount(defaultStreamId), 0);

    // Now we have lost track of the amount owed to the recipient.

    resetPrank({ msgSender: users.sender });
    flow.deposit(defaultStreamId, DEPOSIT_AMOUNT);
    assertGt(flow.getBalance(defaultStreamId), 0);

    // Lets try to withdraw the remaining
    resetPrank({ msgSender: users.recipient });
    flow.withdrawMax({ streamId: defaultStreamId, to: users.recipient });
    assertLt(flow.getBalance(defaultStreamId), DEPOSIT_AMOUNT);
}
```

</details>
